### PR TITLE
gh-98811: use full source location to simplify __future__ imports error checking. This also fixes an incorrect error offset.

### LINKED ISSUE #98811: __futures__ parsing can be simplified using complete source locations for error detection. Also the error's offset is off by one sometimes.
The check for whether there are ``__future__`` imports which are not in the beginning of the script appears in two places:

in [future.c, future_parse](https://github.com/python/cpython/blob/b27b57c6e44a276c8a9843fd37d4cf65b2827d5c/Python/future.c#L57), the ff_lineno is stored (last lineno of a ``__future__`` import) and there is a check for [bad ``__future__`` imports on the same line ](https://github.com/python/cpython/blob/main/Lib/test/badsyntax_future7.py).

Then in [compile.c, compiler_from_import()](https://github.com/python/cpython/blob/b27b57c6e44a276c8a9843fd37d4cf65b2827d5c/Python/compile.c#L3912), any import from ``__future__`` which is in a line after ``c->c_future->ff_lineno`` is rejected.

If we replace ff_lineno by ff_location (the complete location information) then the check in compile.c can use the column info to detect the same-line case, and we no longer need to check for this in future.c. The code of future_parse will become much simpler.

There is also a bug in the future.c exception - the offset is off by 1, and this results in incorrect hilighting of the bad import:

```
python.exe -c "import test.badsyntax_future7.py"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/iritkatriel/src/cpython/Lib/test/badsyntax_future7.py", line 3
    from __future__ import nested_scopes; import string; from __future__ import \
                                                        ^
SyntaxError: from __future__ imports must occur at the beginning of the file
```

My patch will result in this output instead:

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/iritkatriel/src/cpython-654/Lib/test/badsyntax_future7.py", line 3
    from __future__ import nested_scopes; import string; from __future__ import \
                                                         ^^^^^^^^^^^^^^^^^^^^^^^^
SyntaxError: from __future__ imports must occur at the beginning of the file
```

* PR: gh-98812

### PR DESCRIPTION
**Pull request by iritkatriel**
Fixes #98811 .

* Issue: gh-98811

### CODE REVIEW

File: Lib/test/test_future.py [hunk 1]
```diff
@@ -60,7 +60,7 @@ def test_badfuture6(self):
     def test_badfuture7(self):
         with self.assertRaises(SyntaxError) as cm:
             from test import badsyntax_future7
-        self.check_syntax_error(cm.exception, "badsyntax_future7", 3, 53)
+        self.check_syntax_error(cm.exception, "badsyntax_future7", 3, 54)
```
Comment (iritkatriel) [inline Lib/test/test_future.py:63]: 53 was wrong. It caused this:

```
python.exe -c "import test.badsyntax_future7.py"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/iritkatriel/src/cpython/Lib/test/badsyntax_future7.py", line 3
    from __future__ import nested_scopes; import string; from __future__ import \
                                                        ^
SyntaxError: from __future__ imports must occur at the beginning of the file
```

which should be this:

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/iritkatriel/src/cpython-654/Lib/test/badsyntax_future7.py", line 3
    from __future__ import nested_scopes; import string; from __future__ import \
                                                         ^^^^^^^^^^^^^^^^^^^^^^^^
SyntaxError: from __future__ imports must occur at the beginning of the file
```

File: Misc/NEWS.d/next/Core and Builtins/2022-10-28-13-59-51.gh-issue-98811.XQypJa.rst [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Use complete source locations to simplify detection of ``__future__``
+imports which are not at the beginning of the file. Also corrects the offset
+in the exception raised in one case, which was off by one and impeded
+hilighting.
```
Comment (AlexWaygood) [inline Misc/NEWS.d/next/Core and Builtins/2022-10-28-13-59-51.gh-issue-98811.XQypJa.rst]:
```suggestion
highlighting.
```

File: Python/compile.c [hunk 3]
```diff
@@ -125,18 +125,20 @@
...
-    int end_col_offset;
-} location;
+typedef PyCompilerSrcLocation location;
 
 #define LOCATION(LNO, END_LNO, COL, END_COL) \
     ((const location){(LNO), (END_LNO), (COL), (END_COL)})
 
 static location NO_LOCATION = {-1, -1, -1, -1};
 
+#define LOCATION_IS_GT(L1, L2) \
```
Comment (markshannon) [inline Python/compile.c]: Make this an inline function, please.
Also, maybe `compare_location(l1, l2)` for more generality?

**Review by markshannon — APPROVED** [review]
LGTM. Maybe convert a macro to an inline function?

File: Python/compile.c (hunk 3 above)
Comment (iritkatriel) [reply Python/compile.c]: It's not a total order - how would this function work?

Comment (iritkatriel) [reply Python/compile.c]: Could be "lt, gt, or neither", and then you can't use it for equality. So we add an equality testing function too?

Comment (iritkatriel) [reply Python/compile.c]: I'll make a function just for gt for now - checking both directions would be less efficient (if it's not GT we don't care about the other distinction) and we don't have a use case for it now. We can always add a location_compare that does two GT calls if we need it.

Comment (markshannon) [reply Python/compile.c]: > It's not a total order - how would this function work?

Good point.

File: Include/cpython/compile.h [hunk 4]
```diff
@@ -31,11 +31,26 @@ typedef struct {
 #define _PyCompilerFlags_INIT \
     (PyCompilerFlags){.cf_flags = 0, .cf_feature_version = PY_MINOR_VERSION}
 
+/* source location information */
+typedef struct {
+    int lineno;
+    int end_lineno;
+    int col_offset;
+    int end_col_offset;
+} PyCompilerSrcLocation;
```
Comment (iritkatriel) [inline Include/cpython/compile.h]: I wonder if I should put an underscore so it's ``_PyCompilerSrcLocation``. Are we ready to make this public?

**Pull request merged**